### PR TITLE
Feature/add custom element built in support

### LINKED ISSUE #136: Support custom elements V1 extension of built-in elements
The custom elements V1 spec defines that existing (built-in) HTML elements can be extended by `<div is="my-foo">` attribute or [by a second optional paramenter of `document.createElement('div', { is: 'my-foo'})`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement#Syntax)

Same is true for namespaced elements:
https://developer.mozilla.org/en-US/docs/Web/API/Document/createElementNS

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #136 

* adds template literal support for extended built-in elements (custom elements V1 spec)
* browserify and babel transform upgrades

### CODE REVIEW

**Comment by AndyOGo** [comment]
This would still need some tests, though for these we would need some additional requirements like polyfills for custom elements V1

**Comment by AndyOGo** [comment]
@goto-bus-stop 
@bcomnes 
@shama 
@yoshuawuyts 
I have added tests:)
May I asked to review it?

File: tests/browser/elements.js [hunk 1]
```diff
@@ -180,3 +180,26 @@ test('allow objects to be passed', function (t) {
...
 })
+
+test('supports extended build-in elements', function (t) {
+  t.plan(1)
+
+  var originalCreateElement = document.createElement
+  var optionsArg
+
+  // this iife is a must to avoid illegal invocation type errors, caused by transformed nanohtml tests
+  (function () {
```
Comment (AndyOGo) [inline tests/browser/elements.js]: If tape would have `before`/`after` or `setup`/`teardown` support it could be implemented cleaner

**Comment by AndyOGo** [comment]
@goto-bus-stop 
@bcomnes 
@yoshuawuyts 
Hope you have some free time to look into this.
I'm looking forward for your feedback

**Comment by bcomnes** [comment]
I would guess @goto-bus-stop is the most qualified to review this.  I do know if we extend nanohtml we may also need to extend/test yoyoify to support the new changes.

**Comment by goto-bus-stop** [comment]
maybe yo-yoify should be deprecated on npm, this module includes a browserify transform

i had a quick glance earlier and the approach looks good. i can prob review this proper on the weekend

**Comment by bcomnes** [comment]
Oh yeahhhh. I forgot about that. +1 to that

**Comment by AndyOGo** [comment]
@bcomnes 
@goto-bus-stop 
Awesome guys, thanks for your feedback and this great library.
Take your time to review it unhurriedly, I'm looking forward for your review:)

File: lib/browser.js [hunk 2]
```diff
@@ -23,11 +23,24 @@ function nanoHtmlCreateElement (tag, props, children) {
...
-    el = document.createElementNS(ns, tag)
+    if (isCustomElement) {
+      el = document.createElementNS(ns, tag, { is: isCustomElement })
+    } else {
+      el = document.createElementNS(ns, tag)
+    }
   } else if (tag === COMMENT_TAG) {
     return document.createComment(props.comment)
+  } else if (isCustomElement) {
+    el = document.createElement(tag, { is: isCustomElement })
```
Comment (AndyOGo) [inline lib/browser.js]: please see the second optional `options` parameter
https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement#Syntax

File: lib/browser.js [hunk 3]
```diff
@@ -23,11 +23,24 @@ function nanoHtmlCreateElement (tag, props, children) {
...
+  if (props.is) {
+    isCustomElement = props.is
+    delete props.is
+  }
+
   // Create the element
   if (ns) {
-    el = document.createElementNS(ns, tag)
+    if (isCustomElement) {
+      el = document.createElementNS(ns, tag, { is: isCustomElement })
```
Comment (AndyOGo) [inline lib/browser.js]: please see third optional `options` parameter
https://developer.mozilla.org/en-US/docs/Web/API/Document/createElementNS#Syntax

**Comment by AndyOGo** [comment]
Hi @goto-bus-stop 
hmh, I guess you have lots of stuff to do?
Unfortunately I have a giant refactoring waiting to be reviewed and merged but it relies upon this PR.
I hope you are happy with it and I would be super lucky if you find some time soon to check it 🙏

File: tests/browser/elements.js [hunk 4]
```diff
@@ -180,3 +180,26 @@ test('allow objects to be passed', function (t) {
...
+      optionsArg = arguments[1]
+      return originalCreateElement.apply(this, arguments)
+    }
+  })()
+
+  ;html`<div is="my-div"></div>`
+
+  t.ok(typeof optionsArg === 'object' && optionsArg.is === 'my-div', 'properly passes optional extends object')
+
+  delete document.createElement
```
Comment (goto-bus-stop) [inline tests/browser/elements.js]: `document.createElement` should be reset to `originalCreateElement` here i think? else future tests below this one won't be able to use it.

Comment (AndyOGo) [reply tests/browser/elements.js]: @goto-bus-stop 
You absolutely right.
Actaully this is what I want to archieve by `delete document.createElement`. It's a bit opaque though.
But let me explain.
1) the original `createElement` method sits in the [`HTMLDocument.prototype` object](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDocument) of the [Document interface](https://developer.mozilla.org/en-US/docs/Web/API/Document)
2) I cache [`document.createElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement) from there
3) Now here is the important bit, I store a new `createElement` method at the `document` instance directly, shadowing the original at it's prototype below
4) that's why deleting it should be sufficient as I understand it

Are you fine with that?
Else I'm also okay to explicitly reassign the original.

**Comment by AndyOGo** [comment]
Awesome:) I have a little question too:)

File: tests/browser/elements.js (hunk 4 above)
Comment (goto-bus-stop) [reply tests/browser/elements.js]: oh i see, that makes sense. i'd like a comment to clarify that though. something short like `// revert to prototype method` might be enough

Comment (AndyOGo) [reply tests/browser/elements.js]: awesome idea, just did it 💪

**Comment by AndyOGo** [comment]
Awesome, thank you so much :)
Just installed it:)

And it works great.

I'm doing a big refactoring which extends built-in elements.
- Generic Built-in Support https://github.com/axa-ch/patterns-library/pull/642
- Enhanced `<table is="">` built-ins https://github.com/axa-ch/patterns-library/pull/613

Btw. I read in your release notes that *autonomous custom elements* `<custom-element>` aren't still support. I wonder how you mean it, because I use the heavily since months 🤔

**Comment by goto-bus-stop** [comment]
lol, you're right. i was thinking of #121 but that's actually not about `<custom-element>`, but about passing in custom element classes as objects `<${CustomElement} />`. updated the release note, thanks for the correction! :laughing:

**Pull request merged**